### PR TITLE
Add TPC.v4 with quantization preservation.

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
This PR handles quantization preservation in quantization options filtering. For example: in this network:
`add -> reshape -> conv`
the output of the add node can't have a 16 bit activation because the conv node doesn't support 16 bit inputs. reshape node is quantization preserving that supports both 8 and 16 bits.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 1]
```diff
@@ -79,6 +79,68 @@ def set_quantization_configuration_to_graph(graph: Graph,
     return graph
 
 
+def filter_node_qco_by_graph(node: BaseNode, tpc: TargetPlatformCapabilities,
+                             graph: Graph, node_qc_options: QuantizationConfigOptions
+                             ) -> Tuple[OpQuantizationConfig, List[OpQuantizationConfig]]:
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:84]: Please start a new line for each argument, and detail in the comments what it returns.
Also, is it possible to split this function insto smaller functions?

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 2]
```diff
@@ -79,6 +79,68 @@ def set_quantization_configuration_to_graph(graph: Graph,
...
+        next_nodes_qc_options = [_node.get_qco(tpc) for _node in next_nodes]
+        next_nodes_supported_input_bitwidth = min([op_cfg.max_input_activation_n_bits
+                                                   for qc_opts in next_nodes_qc_options
+                                                   for op_cfg in qc_opts.quantization_config_list])
+
+        # Filter node's QC options that match next nodes input bit-width.
+        _node_qc_options = [_option for _option in _node_qc_options
+                            if _option.activation_n_bits <= next_nodes_supported_input_bitwidth]
+        if len(_node_qc_options) == 0:
+            Logger.critical(f"Graph doesn't match TPC bit configurations: {node} -> {next_nodes}.")  # pragma: no cover
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:127]: Why can't we build a test for it instead of the no-cover?

File: model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tp_model.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        # Each OperatorsSet has a unique label.
+        # If a quantization configuration options is passed, these options will
+        # be used for operations that will be attached to this set's label.
+        # Otherwise, it will be a configure-less set (used in fusing):
+
+        generated_tpm.set_simd_padding(is_simd_padding=True)
+
+        # May suit for operations like: Dropout, Reshape, etc.
+        default_qco = tp.get_default_quantization_config_options()
+        tp.OperatorsSet("NoQuantization",
```
Comment (reuvenperetz) [inline model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tp_model.py:198]: Do we still need this "NoQuantization"?

File: model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tpc_keras.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                               Dropout,
+                               MaxPooling2D,
+                               tf.split,
+                               tf.gather,
+                               tf.cast,
+                               tf.unstack,
+                               tf.compat.v1.gather,
+                               tf.__operators__.getitem,
+                               tf.strided_slice]
+    quantization_preserving_list_16bit_input = [Identity,
```
Comment (reuvenperetz) [inline model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tpc_keras.py:83]: Maybe remove the identity? We have a substitution to remove it, so it should not affect it anyway.

File: model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tp_model.py (hunk 3 above)
Comment (elad-c) [reply model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tp_model.py:198]: Yes. We don't handle these nodes' quantization yet.

**Review by reuvenperetz — APPROVED** [review]
